### PR TITLE
GET endpoint for Compact/Jurisdiction Privilege Fees

### LINKED ISSUE #209: Licensee selects privileges to buy
dependent on #151 
- OH, KY, NE as test states in all compacts
- states do not automatically get added to production environment
- endpoint recalls all state-related data
- when a user already has a privilege or a home state license in a state, the privilege option is gray

### PR DESCRIPTION
**Pull request by landonshumway-ia**
As part of this ticket: https://github.com/csg-org/CompactConnect/issues/209 we have a need
for compacts and jurisdictions to define how much they plan to charge for a provider 
to purchase privileges. To save costs on our initial implementation, we will **not** collect this information
through an API endpoint. Instead that information will need to be stored within YAML configuration
files within this project. This commit adds documentation outlining the onboarding process, 
and provides configuration file placeholders

As part of this change, we are introducing a new DynamoDB table to hold compact configuration.
This is because we want to keep our provider information separated from our compact configuration
in terms of where it is stored. In order to upload the YAML configuration in the Dynamo table, we have 
added a CloudFormation custom resource, which is essentially a lambda function that runs at deploy time
to upload this configuration.

Finally, this adds a new GET `/v1/purchases/privileges/options/` endpoint, which will return all the jurisdiction and
compact fee information related to privileges. This will be used by the frontend to display the list of available 
jurisdiction privileges for a user to select from.

The high-level design for the YAML compact configuration files and onboarding process was reviewed and 
approved by @isabeleliassen during a previous coordination meeting. 

### Requirements List
- No new requirements needed to run deployment

### Description List
- Added YAML compact configuration files and onboarding documentation
- Added custom resource to convert YAML to JSON and upload configuration into DynamoDB
- Updated CDK tests to leverage snapshots for large template comparisons.
- Added GET `/v1/purchases/privileges/options/` endpoint in new `purchases` directory.

### Testing List
- `yarn test:unit:all` should run without errors or warnings
- `yarn serve` should run without errors or warnings
- `yarn build` should run without errors or warnings
- For API configuration changes: CDK tests added/updated in `backend/compact-connect/tests/unit/test_api.py`
- Code review

Closes #209 
(backend work only, frontend will be done in separate PR)

### CODE REVIEW

**Comment by jlkravitz** [comment]
- [ ] review the pull request to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] skim all new code, in the context of existing code, looking for problems (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] methodically review all new tests for correctness, quality of naming
	- [ ] look at code coverage of tests
	- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] run a security audit of dependencies (e.g. `npm audit` and `pip audit`) to ensure that there are no vulnerabilities that will be deployed to production (as opposed to vulnerabilities that only have an impact on the development environment)

File: backend/compact-connect/lambdas/custom-resources/tests/function/test_handlers/test_compact_configuration_uploader.py
Comment (jlkravitz) [inline backend/compact-connect/lambdas/custom-resources/tests/function/test_handlers/test_compact_configuration_uploader.py:48]: implication being? a comparison would fail since this function might return different dates during the same run of a test?

File: backend/compact-connect/lambdas/purchases/config.py
Comment (jlkravitz) [inline backend/compact-connect/lambdas/purchases/config.py:15]: Where is this used, and for what purpose?

File: backend/compact-connect/lambdas/purchases/data_model/schema/compact.py
Comment (jlkravitz) [inline backend/compact-connect/lambdas/purchases/data_model/schema/compact.py:24]: Looks like this is _not_ required then? If that's true, can we update the README?

File: backend/compact-connect/lambdas/purchases/data_model/schema/jurisdiction.py
Comment (jlkravitz) [inline backend/compact-connect/lambdas/purchases/data_model/schema/jurisdiction.py:31]: Can we make clear in the docs that this nested config is not required, or otherwise make this required?

File: backend/compact-connect/lambdas/purchases/tests/function/__init__.py
Comment (jlkravitz) [inline backend/compact-connect/lambdas/purchases/tests/function/__init__.py:77]: Isn't this the purchases lambda? Why would the type ever be provider?

File: backend/compact-connect/lambdas/purchases/tests/function/test_handlers/test_privilege_options.py
Comment (jlkravitz) [inline backend/compact-connect/lambdas/purchases/tests/function/test_handlers/test_privilege_options.py:43]: should we test that this list is only of length 1 before pulling out the first element?

File: backend/compact-connect/lambdas/purchases/tests/function/test_handlers/test_privilege_options.py
Comment (jlkravitz) [inline backend/compact-connect/lambdas/purchases/tests/function/test_handlers/test_privilege_options.py:69]: should we test that this list is only of length 1 before pulling out the first element?

File: backend/compact-connect/stacks/persistent_stack/compact_configuration_upload.py
Comment (jlkravitz) [inline backend/compact-connect/stacks/persistent_stack/compact_configuration_upload.py:170]: Is there a way to 'test' this locally? e.g., I can imagine misspelling a compact directory and not noticing, and it might be nice if a test would print the active compact and jurisdiction configurations under a particular environment.

File: backend/compact-connect/tests/unit/test_api.py
Comment (jlkravitz) [inline backend/compact-connect/tests/unit/test_api.py:77]: What's this?

File: backend/compact-connect/tests/unit/test_api.py
Comment (jlkravitz) [inline backend/compact-connect/tests/unit/test_api.py:58]: Oh, this might be your solution to my prior comment re "testing." Let me know if that's right.

Comment (jlkravitz) [reply backend/compact-connect/tests/unit/test_api.py:58]: Also, looks like this function is only needed in the `test_app.py` file, not here?

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (jlkravitz) [inline backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:1]: # General Comments

* This looks great! Thanks for the thorough documentation- really helped me walk through and understand the codebase. I also appreciate your detailed commit messages, which answered some of my questions before I even walked through the code.
* I noticed we're copying a lot of the code for the lambda (e.g., the pagination and api_handler code and their corresponding tests). I don't see this as a huge issue if we only do so once or twice, but is there a plan to constrain the duplication going forward? Perhaps you don't expect to need to make similar lambdas again, or there are longer-term strategies for sharing code. No action needed on this PR, but I just wanted to throw it out there to make sure it's something we're thinking about.

**Review by jlkravitz — CHANGES_REQUESTED** [review]

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (landonshumway-ia) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:69]: @jlkravitz I would need this information from management. For this current effort, I was asked to just put up the initial document without specific details on future implementation.

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (landonshumway-ia) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:83]: Yes, the intended audience for this document are technical maintainers and IT staff that would be creating the new configuration files whenever a jurisdiction is onboarded into the system. I'll add the link in an upcoming commit.

File: backend/compact-connect/compact-config/aslp/kentucky.yml
Comment (landonshumway-ia) [reply backend/compact-connect/compact-config/aslp/kentucky.yml:1]: I would not recommend it. The issue this is addressing is we are being asked to deliver a service for which there is currently no active jurisdictions data. In other words, we are trying to put the cart before the horse and we have to have configuration files in order to show something to our stakeholders before they give us real data. Eventually, these files will hold real values for these jurisdictions. Any effort we put into making different sub/test directories will just be throw away work.

Comment (landonshumway-ia) [reply backend/compact-connect/compact-config/aslp/kentucky.yml:1]: I can reword the comment to make it more apparent that this is placeholder data. Not necessarily just for testing, but it is also not the actual configuration for these states.

File: backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py:59]: If the compact or jurisdiction is not part of the activeEnvironment, it is not included in the input sent to the custom resource, so it will not be loaded.

File: backend/compact-connect/stacks/persistent_stack/compact_configuration_upload.py
Comment (landonshumway-ia) [reply backend/compact-connect/stacks/persistent_stack/compact_configuration_upload.py:170]: The way you would test this locally is by running the CDK synth for the persistent stack and then checking the CFN template that is generated. The template will include the JSON formatted string that is generated and being sent to the custom resource.

File: backend/compact-connect/tests/unit/test_api.py
Comment (landonshumway-ia) [reply backend/compact-connect/tests/unit/test_api.py:77]: This is used to show large diffs when running unittest's `assertEqual()` method. Because the snapshots we are checking can be fairly large, if we don't set this here, the developer will not see the diff between the snapshot and the actual value which caused the test to fail.

File: backend/compact-connect/tests/unit/test_api.py
Comment (landonshumway-ia) [reply backend/compact-connect/tests/unit/test_api.py:58]: Yes, this test is how we confirm that the compact configuration matches expected values for the test environment. This code is used by both the `test_app` and the child classes of `test_api` [here](https://github.com/csg-org/CompactConnect/pull/243/files#diff-9dfe2171d4a0ef351c005ebda8009b928b8a129478a27d0aee0c341ecf11c1feR93-R95) and [here](https://github.com/csg-org/CompactConnect/pull/243/files#diff-9ab2f250e93c65ce67302cb35ebd8f91210a98ab3335696ac494f740f2f07304R109-R110)

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (landonshumway-ia) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:42]: Unfortunately, I don't have that information currently 😞 . I agree it would be nice to have that information, but until we have a live system with actual use cases for operation support, it really could be either, but any attempt we try to pin down today is going to eventually be updated. If this is to be regarded as a living document, which we actively update and maintain as part of adding new features, then those details will come as we have a better understanding. That is the meaning of agile, right? This is a first iteration of a living document. If you want to flesh this out with a waterfall approach, we can go get every business detail ironed out so that it can be captured in the document today, and we'll still have to update it when management changes requirements on us next month 😄 

So again, I agree, more fleshed out detail will be great. When we reach the point of implementation that gives us a better picture of how the system will send these types of notifications, it can be added to the document.

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (landonshumway-ia) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:56]: see comment above

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (landonshumway-ia) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:154]: The intended audience is any technical developer that maintains / onboards new compacts/jurisdictions into the system using these YAML files. So we need this section to explain the purpose of these top level compact files and how they are configured.

File: backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py:32]: Yes, in the case of delete we do not remove any data from the table, this will be handled by the DB's removal policy. If we retain the data for the DB, we don't want to delete anything here. If the removal policy deletes everything, then there is no need for us to duplicate that effort here.

File: backend/compact-connect/lambdas/purchases/tests/function/test_handlers/test_privilege_options.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/purchases/tests/function/test_handlers/test_privilege_options.py:69]: We can, either way the test would fail, since it will raise an index error and the developer would need to correct the issue regardless.

File: backend/compact-connect/tests/unit/test_api.py
Comment (landonshumway-ia) [reply backend/compact-connect/tests/unit/test_api.py:58]: It's worth pointing out that @jusdino will be refactoring all of these tests in an upcoming PR to reduce duplicating stack synth steps and make them easier to maintain in general.

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (landonshumway-ia) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md]: I've added documentation to this section to better explain these details.

File: backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py:32]: I'll add a comment to the code to denote this intended behavior

File: backend/compact-connect/lambdas/purchases/config.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/purchases/config.py:15]: good catch - copy/paste from other directory. I'll remove it.

File: backend/compact-connect/lambdas/purchases/data_model/schema/compact.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/purchases/data_model/schema/compact.py:24]: Good catch. I initially left many of these fields optional as I was adding tests for the endpoint, but this should definitely be required.

File: backend/compact-connect/lambdas/purchases/data_model/schema/jurisdiction.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/purchases/data_model/schema/jurisdiction.py:31]: I've updated the docs to show that this field is optional

File: backend/compact-connect/lambdas/purchases/tests/function/__init__.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/purchases/tests/function/__init__.py:77]: good catch 👍, copy/paste from previous test directory

File: backend/compact-connect/lambdas/purchases/tests/function/test_handlers/test_privilege_options.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/purchases/tests/function/test_handlers/test_privilege_options.py:69]: Ah never mind, I misread your comment. Sure I'll add that check in.

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (jlkravitz) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:45]: I by no means am implying we should use a waterfall approach :')

That makes sense! I mostly wanted to clarify if the detail was intentionally sparse.

I guess one thought, and this is truly meant to be a question for discussion (I'm not an expert on this): Is including these configuration fields antithetical to the concept of "vertical slicing"? It seems like we're adding configuration for something that does...nothing. Like adding a backend component for a non-existent frontend.

Consider this a rhetorical question for the sake of this PR (i.e., one we don't have to discuss back-and-forth here or resolve to merge), but I thought it might be interesting to discuss.

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (jlkravitz) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:72]: Got it! That's fine.

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (jlkravitz) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:162]: Got it, thank you! (And thanks for bearing with my questions here.)

File: backend/compact-connect/compact-config/aslp/kentucky.yml
Comment (jlkravitz) [reply backend/compact-connect/compact-config/aslp/kentucky.yml]: Ahh, I got it. I misunderstood. Calling it placeholder data makes sense to me!

File: backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py
Comment (jlkravitz) [reply backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py:59]: Ah got it. Thanks!

File: backend/compact-connect/lambdas/purchases/data_model/schema/jurisdiction.py
Comment (jlkravitz) [reply backend/compact-connect/lambdas/purchases/data_model/schema/jurisdiction.py:31]: Thank you!

File: backend/compact-connect/stacks/persistent_stack/compact_configuration_upload.py
Comment (jlkravitz) [reply backend/compact-connect/stacks/persistent_stack/compact_configuration_upload.py:170]: Got it. Helpful! Is that worth documenting in the markdown file above? (It might not be given the intended audience, genuine question. Your call!)

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (landonshumway-ia) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:45]: That is a legitimate question, and I'd be curious to get @jusdino's thoughts on it. The initial requirements given to us as part of designing this was to have _something_ in place for email notifications for the three use cases outlined in the document, so I threw them in essentially as a first attempt given we know there _will_ be notifications for these elements, and that's about it 😆 I know we have stories coming up on the backlog for adding email notifications, so the fields will likely be utilized sooner rather than later.

But yeah honestly, in the interest of vertical slicing we could have added them in later, something to consider during our next retrospective.

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (landonshumway-ia) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md]: I also added towards the top of the document denoting who the intended audience is for these onboarding steps.

File: backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py]: Good catch. I've removed them.

File: backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/custom-resources/handlers/compact_config_uploader.py]: Removed unused params from docs

File: backend/compact-connect/lambdas/custom-resources/tests/function/test_handlers/test_compact_configuration_uploader.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/custom-resources/tests/function/test_handlers/test_compact_configuration_uploader.py]: Yes, I've updated the comment to make this more explicit

File: backend/compact-connect/lambdas/purchases/tests/function/test_handlers/test_privilege_options.py
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/purchases/tests/function/test_handlers/test_privilege_options.py]: added that check

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (landonshumway-ia) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:1]: Great points, I discussed this briefly with @jusdino last week and we agree the duplicated utility, config, and testing code should be addressed if we add more directories like this. The way I've tackled this before is to add a common directory with the common modules inside, and deploy the common code as a lambda layer which is then shared across all the lambdas in the CDK app. There are of course other solutions we can consider but that is one of the simplest.

So yeah, we should definitely discuss during our next spring retro on how we would want to address common code for the project.

File: backend/compact-connect/stacks/persistent_stack/compact_configuration_upload.py
Comment (landonshumway-ia) [reply backend/compact-connect/stacks/persistent_stack/compact_configuration_upload.py:170]: That's a good question... it might be helpful to briefly mention that making any changes to the files may break the snapshot tests. So yes, I'll add a brief testing section in there.

File: backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md
Comment (jlkravitz) [reply backend/compact-connect/docs/onboarding/JURISDICTION_COMPACT_ONBOARDING.md:1]: Awesome. That makes a lot of sense, and we can cross that bridge when we get there.

**Review by jlkravitz — APPROVED** [review]
@isabeleliassen good to merge!

**Review by carlsims — APPROVED** [review]

**Pull request merged**
